pkg/image/registryclient/client.go

```diff
@@ -374,6 +426,24 @@ func NewLimitedRetryRepository(repo distribution.Repository, retries int, limite
 	}
 }
 
+func (c *Context) reposForRef(ctx context.Context, locator imagereference.DockerImageReference, sources []imagereference.DockerImageReference) error {
+	cachedRepos := make(map[imagereference.DockerImageReference]distribution.Repository)
+	for _, altRef := range sources {
+		altRepo, err := c.Repository(ctx, altRef.DockerClientDefaults().RegistryURL(), altRef.RepositoryName(), true)
+		if err != nil {
+			return err
+		}
+		if altRepo != nil {
+			cachedRepos[altRef] = altRepo
+		}
+	}
+	if len(cachedRepos) == 0 {
+		return fmt.Errorf("did not find any repositories for %s", locator)
+	}
+	c.cachedRepos = cachedRepos
+	return nil
+}
+
 // isTemporaryHTTPError returns true if the error indicates a temporary or partial HTTP failure
 func isTemporaryHTTPError(err error) (time.Duration, bool) {
```

Comment:
Was expecting to see one example of this in use with the blob service / manifest service wrapper (the Get method) showing how you will use these methods

Reply:
You mean a method that does more than return the first repo and ref? `oc` uses the `retryRepository` - I found it isn't necessary to create a new repository wrapper.  and the manifests.Get is called in the strategy method, `RepositoryForAlternateReference`.  I've only implemented the`simpleLookupICSP` strategy that `oc` will use.   
I added the Context method `NextRepositoryWithAlternateReference` to return the next (only&preferred for oc) repo, ref found from RepositoryForAlternateReference. Depending on the strategy,  RepoForAlternateRef may have more complex ordering, another strategy can be used but I only added the simpleLookupICSP to get the next 'successful' repo, ref. 
I may be misunderstanding your comment.

Reply:
No one in this PR calls NextRepositoryWithAlternateReference, which means someone outside this library has to interpret it.  I was expecting `retryRepository` to invoke alternate lookups instead - so that someone outside this library doesn't (and I could see the structure change for retryRepository.

Reply:
I've updated, `nextRepositoryWithReference` is retryRepository method, and 

> I have coded an example of usage of this change in [here](https://github.com/ricardomaraschini/dump-place/blob/master/pull-through-mirror/icspaware.go#L80-L134). What I noticed is that it could be a little bit easier if `NextRepositoryWithAlternateReference` returned the original repository as the last call. If it did that [this](https://github.com/ricardomaraschini/dump-place/blob/master/pull-through-mirror/icspaware.go#L119-L133) would not be needed.

I reworked this based on @smarterclayton  comment - the exported Context method `RepositoryWithAlternateReference` returns the first alternate repo from the retryRepository method `nextRepositoryForAlternateReference`  and the Context.cachedRepos is updated accordingly. The exported `RepositoryWithAlternateReference` only returns repos for which the ManifestService.Get(ctx, locator-digest) succeeds. The caller can collect those and do more ordering if necessary.

Reply:
^^ outdated ^^ removed the cachedRepos